fix: return non-secure cookies with HTTPS URLs

zevisert opened the pull request:
Cookies have a "Secure" attribute which tells the browsers that a given cookie should only be sent via HTTPS. In it's absense "Secure" is falsy and these cookies should be sent with both HTTP and HTTPS requests. Playwright now returns only the "Non-Secure" cookies for HTTP URLs, and both "Secure" and "Non-Secure" cookies for HTTPS URLs.

Fixes #5504

zevisert committed:
fix: return non-secure cookies with HTTPS URLs

Cookies have a "Secure" attribute which tells the browsers
that a given cookie should only be sent via HTTPS. In it's
absense "Secure" is falsy and these cookies should be sent
with both HTTP and HTTPS requests. Playwright now returns
only the "Non-Secure" cookies for HTTP URLs, and both
"Secure" and "Non-Secure" cookies for HTTPS URLs.

Fixes #5504

zevisert commented:
I didn't end up changing the API docs for this fix; under `BrowserContext.cookies` we currently have

> If no URLs are specified, this method returns all cookies. If URLs are specified, only cookies that affect those URLs are returned.

And that seems appropriate still. I thought about adding another sentence there, along the lines of...
> If no URLs are specified, this method returns all cookies. If URLs are specified, only cookies that affect those URLs are returned. Secure cookies are only returned for HTTPS URLs, while non-secure cookies are returned for both HTTP and HTTPS URLs.

But that seems like part of "only cookies that affect those URLs". Let me know what you think!

mxschmitt commented on test/browsercontext-cookies.spec.ts:
```diff
@@ -210,3 +210,46 @@ it('should not return cookies with empty value', async ({context, page, server})
+    secure: false
+  }]);
+  const cookies = await context.cookies('https://foo.com');
+  expect(cookies.sort((a, b) => a.name.localeCompare(b.name))).toEqual([{
```
```suggestion
  expect(new Set([cookies])).toEqual(new Set([{
```
If I get you right, you want to have a sort independent toEqual, then you can use a Set for that.

zevisert replied:
Ah yeah, good idea. Two of the other tests in the file also use a `.sort`, should I change those here too? I'd make this change  locally anyway, since your suggestion doesn't have a new closing paren for the second `Set` and `cookies` is already an iterable.

zevisert replied:
I changed them all.

zevisert committed:
chore: use sets for group equality in context text

Some tests in //tests/browsercontext-cookies.spec.ts
used a mix of Array.sort and str.localeCompare. This
comes from a suggestion from @mxschmitt.

See #5507

zevisert commented:
Oh darn, that last commit should have read `chore: use sets for group equality in cookie tests`, not `in context text` 🤦. If it's a big deal, let me know and I can rebase

aslushnikov commented on src/server/network.ts:
```diff
@@ -36,7 +36,7 @@ export function filterCookies(cookies: types.NetworkCookie[], urls: string[]): t
       if (!parsedURL.pathname.startsWith(c.path))
         continue;
-      if ((parsedURL.protocol === 'https:') !== c.secure)
+      if ((parsedURL.protocol !== 'https:') && c.secure)
```
nit: drop parenthesis 

```suggestion
      if (parsedURL.protocol !== 'https:' && c.secure)
```

aslushnikov committed:
Update src/server/network.ts
